File: demo/bench.mjs
// demo/bench.mjs <wasm> <fixture> — time the REAL Req pipeline on WasmGC, network-free.
// The fetch already happened once (captured to the fixture); the host http.get just hands back those
// in-memory bytes, so e.run() is pure compute: Req.new/merge → URI.parse → step pipeline → decode_body →
// extract JS URLs → :crypto.hash(:sha256). Report module-compile, instantiate, and per-run cost.
import fs from "node:fs";
import nodeCrypto from "node:crypto";
import { makeBig, makeMath, makeStr, makeCrypto, makeProcStubs, binCodec } from "../runtime/imports.mjs";
const [wasmPath, fixturePath] = process.argv.slice(2);
const fixture = fs.readFileSync(fixturePath);
const bytes = fs.readFileSync(wasmPath);

const big = makeBig();
const math = makeMath();
let e;
const { wrBytes } = binCodec(() => e);
const str = makeStr(() => e);
// the response body lives in wasm memory (built ONCE); http.get returns that ref — so we measure the Req
// pipeline + parse + hash, not the byte-by-byte boundary copy. (A real host would bulk-copy once/request.)
let fixtureBin = null;
const http = { get: _ => fixtureBin };
const crypto = makeCrypto(() => e, nodeCrypto);
const { proc, sched } = makeProcStubs();

const ms = () => performance.now();
let t = ms(); const mod = new WebAssembly.Module(bytes); const tCompile = ms() - t;
t = ms(); e = new WebAssembly.Instance(mod, { big, math, str, http, crypto, proc, sched }).exports; const tInstance = ms() - t;
fixtureBin = wrBytes(fixture);                      // build the response binary in wasm ONCE

for (let i = 0; i < 50; i++) e.run();              // warm up (JIT)
const N = 5000;
t = ms(); for (let i = 0; i < N; i++) e.run(); const total = ms() - t;

console.log(JSON.stringify({
  wasm_bytes: bytes.length,
  module_compile_ms: +tCompile.toFixed(2),
  instantiate_ms: +tInstance.toFixed(2),
  runs: N,
  per_run_us: +((total / N) * 1000).toFixed(2),
  runs_per_sec: Math.round(N / (total / 1000)),
}));

File: demo/bench.exs
#!/usr/bin/env elixir
# Benchmark the REAL Req program on both engines, NETWORK-FREE (the fetch was captured once; both replay
# the same in-memory bytes). Reuses demo/_work/real.wasm + the fixture from a prior `elixir demo/run_real.exs`.
Mix.install([{:req, "~> 0.5"}])

Code.require_file("../tooling.exs", __DIR__)

defmodule Bench do
  @here Path.dirname(__ENV__.file)
  @node Tooling.node!()

  def main do
    wasm = Path.join(@here, "_work/real.wasm")
    fixture = Path.join(@here, "_work/resy_body.html")
    unless File.exists?(wasm) and File.exists?(fixture),
      do: (IO.puts("run `elixir demo/run_real.exs` first (builds real.wasm + captures the fixture)."); System.halt(1))

    # ── WasmGC ──
    {json, 0} = System.cmd(@node, [Path.join(@here, "bench.mjs"), wasm, fixture])
    w = Jason.decode!(json)

    # ── Elixir VM (real Req, adapter stubbed to the same bytes; no socket) ──
    stub = struct(Req.Response, status: 200, headers: %{"content-type" => ["text/html; charset=utf-8"]}, body: File.read!(fixture))
    Req.default_options(adapter: fn req -> {req, stub} end)
    [{Resy, _} | _] = Code.compile_file(Path.join(@here, "resy.exs")) |> Enum.filter(&(elem(&1, 0) == Resy))
    for _ <- 1..200, do: Resy.run()                       # warm
    n = 5000
    {us, _} = :timer.tc(fn -> for _ <- 1..n, do: Resy.run() end)
    vm_per_run_us = Float.round(us / n, 2)

    IO.puts("\n══════ REAL Req — network-free compute (#{w["runs"]} runs each) ══════\n")
    IO.puts("  wasm module size      #{Float.round(w["wasm_bytes"] / 1024, 0) |> trunc()} KB (unoptimized, -g)")
    IO.puts("  WASM compile+instantiate (one-time)  #{w["module_compile_ms"]} + #{w["instantiate_ms"]} ms")
    IO.puts(String.duplicate("  ─", 28))
    IO.puts("  WasmGC  per run   #{pad(w["per_run_us"])} µs   (#{w["runs_per_sec"]} runs/s)")
    IO.puts("  BEAM    per run   #{pad(vm_per_run_us)} µs   (#{round(1_000_000 / vm_per_run_us)} runs/s)")
    ratio = Float.round(w["per_run_us"] / vm_per_run_us, 2)
    IO.puts("\n  WasmGC is #{ratio}× the BEAM's time#{if ratio < 1, do: "  (faster!)", else: ""}  — same real Req pipeline, no network.")
  end

  defp pad(x), do: String.pad_leading("#{x}", 8)
end

Bench.main()
